Sources/MongoSwift/BSON/Codec.swift

```diff
@@ -333,3 +333,580 @@ private func getBsonValue(_ value: Any) -> BsonValue? {
+        }
+
+        guard let topContainer = self.storage.topContainer as? Document else {
+            throw DecodingError._typeMismatch(at: self.codingPath, expectation: Document.self, reality: self.storage.topContainer)
+        }
+
+        let container = _BsonKeyedDecodingContainer<Key>(referencing: self, wrapping: topContainer)
+        return KeyedDecodingContainer(container)
+    }
+
+    // Returns the data stored in this decoder as represented in a container appropriate for holding a single primitive value.
+    public func singleValueContainer() throws -> SingleValueDecodingContainer {
+        return self
+    }
+
+    // Returns the data stored in this decoder as represented in a container appropriate for holding values with no keys.
+    public func unkeyedContainer() throws -> UnkeyedDecodingContainer {
+        guard self.storage.topContainer != nil else {
+            throw DecodingError.valueNotFound(UnkeyedDecodingContainer.self,
+                                  DecodingError.Context(codingPath: self.codingPath,
+                                                        debugDescription: "Cannot get unkeyed decoding container -- found null value instead."))
+        }
+
+        guard let arr = self.storage.topContainer as? [BsonValue?] else {
+            throw DecodingError._typeMismatch(at: self.codingPath, expectation: [BsonValue?].self, reality: self.storage.topContainer)
+        }
+
+        return _BsonUnkeyedDecodingContainer(referencing: self, wrapping: arr)
+    }
+}
+
+// Storage for a _BsonDecoder.
+private struct _BsonDecodingStorage {
+
+    /// The container stack, consisting of `BsonValue?`s. 
+    private(set) fileprivate var containers: [BsonValue?] = []
+
+    /// Initializes `self` with no containers.
+    fileprivate init() {}
+
+    /// The count of containers stored.
+    fileprivate var count: Int { return self.containers.count }
+
+    /// The container at the top of the stack.
+    fileprivate var topContainer: BsonValue? {
+        precondition(self.containers.count > 0, "Empty container stack.")
+        return self.containers.last!
+    }
+
+    /// Adds a new container to the stack.
+    fileprivate mutating func push(container: BsonValue?) {
+        self.containers.append(container)
+    }
+
+    /// Pops the top container from the stack. 
+    fileprivate mutating func popContainer() {
+        precondition(self.containers.count > 0, "Empty container stack.")
+        self.containers.removeLast()
+    }
+}
+
+/// Just a temporary error to make code compile until we implement everything.
+struct UnimplementedError: LocalizedError {
+    public var errorDescription: String? { return "Unimplemented" }
+}
+
+/// A protocol for types that are not BSON types but require decoding support. 
+private protocol NonBsonValue {
+    init(from: BsonValue) throws
+}
+
+extension Int8: NonBsonValue {
+    init(from: BsonValue) throws {
```

Comment:
Not sure how to directly comment on your proposal in the description, but these stub implementations seem to be the proposal realized? Either way, I agree with the approach laid out there!

Reply:
yup, put them here so things would compile but wanted to hold off on implementing til we discussed - sounds good